tests/benchmark/crud_test.go

```diff
@@ -0,0 +1,841 @@
+// Copyright 2022 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package benchmark
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/apigee/registry/connection"
+	"github.com/apigee/registry/rpc"
+	"google.golang.org/genproto/protobuf/field_mask"
+)
+
+var (
+	registryProject    string
+	versionCount       = 1
+	apiNamePrefix      = "test"
+	apiNameStartOffset = 0
+	specPath           string
+	mimeType           string
+	pauseDuration      int64
+	pageSize           int
+)
+
+func init() {
+	flag.StringVar(&registryProject, "registry_project", "bench", "Name of the Registry project")
+	flag.IntVar(&versionCount, "version_count", 1, "Number of versions of the API to create (Each version will have 1 spec)")
+	flag.StringVar(&apiNamePrefix, "api_prefix", "test", "Prefix to use for the APIs")
+	flag.IntVar(&apiNameStartOffset, "api_offset", 0, "Offset the name of the API with this value")
+	flag.StringVar(&specPath, "spec_path", filepath.Join("testdata", "openapi.yaml"), "Absolute path to the specification file to use")
+	flag.StringVar(&mimeType, "mime_type", "application/x.openapi;version=3.0.0", "MIME type of the spec file")
+	flag.Int64Var(&pauseDuration, "pause_duration", 5, "Number of seconds to pause after every 1000 runs")
+	flag.IntVar(&pageSize, "page_size", 20, "Page size for the List apis")
+}
+
+func getRootResource(b *testing.B) string {
+	b.Helper()
+	return fmt.Sprintf("projects/%s/locations/global", registryProject)
+}
+
+func getApiName(i int, b *testing.B) string {
+	b.Helper()
+	return fmt.Sprintf("%s-%d", apiNamePrefix, i+apiNameStartOffset)
+}
+
+func pauseAfter1000Iterations(i int, b *testing.B) {
+	b.Helper()
+	if i%1000 == 0 {
+		/*
+			Pause 5 seconds every 1000 iterations to reduce chances of failure on
+			CREATE, UPDATE, DELETE
+		*/
+		time.Sleep(time.Duration(pauseDuration * time.Second.Nanoseconds()))
+	}
+}
```

Comment:
These pauses make me uncomfortable because they suggest that we're accepting and working around some odd performance that we don't fully understand (and might need to fix).